File: www/wiki/js/app/helper/md/mdconf.lua

local function console(obj, out)
	out = out or print

	local outlist = {}
	function _print(obj, level, flag)
		-- 避免循环输出
		local obj_str = tostring(obj)
		for _, str in ipairs(outlist) do
			if str == obj_str then
				return
			end
		end
		outlist[#outlist+1] = obj_str

		level = level or 0
		local indent_str = ""
		for i = 1, level do
		  indent_str = indent_str.."    "
		end
	  
		if not flag then
			out(indent_str.."{")
		end
	  
		for k,v in pairs(obj) do
			if type(v) == "table" then 
				out(string.format("%s    %s = {", indent_str, tostring(k)))
				_print(v, level + 1, true)
			elseif type(v) == "string" then
				out(string.format('%s    %s = "%s"', indent_str, tostring(k), tostring(v)))
			else
				out(string.format("%s    %s = %s", indent_str, tostring(k), tostring(v)))
			end
		end
		out(indent_str.."}")
	end
	
	if type(obj) == "table" then
		_print(obj)
	elseif type(obj) == "string" then
		out('"' .. obj .. '"')
	else
		out(tostring(obj))
	end
end

local strings = require("md/strings")

local mdconf = {}

local function confConvert(c) 
	if type(c) ~= "table" then
		return c
	end

	local nc = {}

	if c.length then
		for i = 0, c.length - 1  do
			nc[#nc+1] = confConvert(c[tostring(i)])
		end
	else 
		for key, value in pairs(c) do
			nc[key] = confConvert(c[key])
		end
	end

	return nc
end

-- md 转json对象
mdconf.mdToJson = function(text) 
	local temp_lines = strings(strings(text):trim()):split("\n")
	local lines = {}
	local line = ""
	local conf = {}
	local curConf = conf

	local getObj = function(key) 
		if not key or key == "" then
			return conf
		end
		local keys = strings(key):split(".")
		local tmpConf = conf
		for i, v in ipairs(keys) do
			tmpConf[v] = tmpConf[v] or {}

			local vn = tonumber(v)
			if vn then
				tmpConf.length = tmpConf.length or -1
				if tmpConf.length <= vn then
					tmpConf.length = vn + 1
				end
			end
			tmpConf = tmpConf[v]
		end

		return tmpConf
	end


	local _mdToJson = function(line) 
		local flag, content = line:match('^([-+#]) (.*)')
		local key, value, temp

		content = strings.trim(content)
		if flag == "#" then
			curConf = getObj(content)
		end

		if flag == "+" or flag == "-"  then
			temp = strings.indexOf(content, ":")
			if (temp and temp > 1) then
				key = strings(strings(content):substring(0, temp - 1)):trim() 
				value = strings(strings(content):substring(temp + 1)):trim()
			else 
				curConf.length = curConf.length or 0
				key = tostring(curConf.length)
				value = tostring(strings.trim(content))
				curConf.length = curConf.length + 1
			end

			if value == "true" then
				value = true
			elseif value == "false" then
				value = false
			else 
				value = value
			end

			curConf[key] = value
		end
	end

	local is_comment = false
	for i, l in ipairs(temp_lines) do
		if l:match('^<!--.*-->%s*$') then
		elseif l:match('^<!--') then
			is_comment = true
		elseif is_comment then
			if l:match('-->%s*$') then
				is_comment = false
			end
		elseif not l:match('^[-+#] .*') then
			line = line .. l  .. "\n"
		else
			if line and line ~= "" then
				lines[#lines+1] = line
			end
			line = l
		end
	end

	if line and line ~= "" then
		lines[#lines+1] = line
	end

	if #lines == 1 and not lines[1]:match('^[-+#] .*') then
		return strings(lines[1]):trim()
	 else 
		for _, line in ipairs(lines) do 
			_mdToJson(strings(line):trim())
		end
	end
	
	return confConvert(conf)
end

local function _jsonToMd(obj, key_prefix) 
	local text = ""
	if key_prefix and key_prefix ~= "" then
		key_prefix = key_prefix .. "."
	else 
		key_prefix = ""
	end
	if #obj == 0 then
		-- 单对象对应列表
		for key, value in pairs(obj) do 
			-- 优先写非对象值
			if strings(key):indexOf("$$") ~= 0 and type(value) ~= "table" then
				text = text .. "- " .. key .. " : " .. value .. "\n"
			end
		end
		for key, value in pairs(obj) do 
			-- 写对象值
			if strings(key):indexOf("$$") ~= 0 and type(value) == "table" then
				text = text .. "\n# " .. key_prefix .. key .. "\n"
				text = text .. _jsonToMd(value, key_prefix .. key)
			end
		end
	 else 
		for i, value in ipairs(obj) do
			if type(value) ~= "table" then
				text = text .. "- " .. obj[i] .. "\n"
			else
				text = text .. "\n# " .. key_prefix .. (i-1) .. "\n"
				text = text .. _jsonToMd(value, key_prefix .. (i-1))

			end 
		end
	end

	return text
end

-- json对象转markdown文本
mdconf.jsonToMd = function(obj) 
	-- 非对象直接写入
	if type(obj) ~= "table" then
		return obj
	end
	

	return _jsonToMd(obj)
end

mdconf.test = function() 
	local obj, text
	obj = "hello world"
	text = mdconf.jsonToMd(obj)
	console(text)
	--console(angular.toJson(mdconf.mdToJson(text)))
	console(mdconf.mdToJson(text))

	obj = {key="value"}
	text = mdconf.jsonToMd(obj)
	console(text)
	--console(angular.toJson(mdconf.mdToJson(text)))
	console(mdconf.mdToJson(text))

	obj = {"list1", "list2"}
	text = mdconf.jsonToMd(obj)
	console(text)
	--console(angular.toJson(mdconf.mdToJson(text)))
	console(mdconf.mdToJson(text))

	obj = {{"list1", "list2"}, {"list3", "list4"}}
	text = mdconf.jsonToMd(obj)
	console(text)
	--console(angular.toJson(mdconf.mdToJson(text)))
	console(mdconf.mdToJson(text))

	obj = {key="value", list={"list1", "list2"}}
	text = mdconf.jsonToMd(obj)
	console(text)
	--console(angular.toJson(mdconf.mdToJson(text)))
	console(mdconf.mdToJson(text))

	obj = {key="value", list={{key1="value1"},{key2="value2"}}}
	text = mdconf.jsonToMd(obj)
	console(text)
	--console(angular.toJson(mdconf.mdToJson(text)))
	console(mdconf.mdToJson(text))

	obj = {key="value", list={{key1="value1"},{key2="value2", list={"list1"}}}}
	text = mdconf.jsonToMd(obj)
	console(text)
	--console(angular.toJson(mdconf.mdToJson(text)))
	obj = mdconf.mdToJson(text)
	console(obj)
end

--mdconf.test()

return mdconf

File: www/wiki/js/app/helper/md/mdconf.js

define([
], function () {
	var mdconf = {};

	// md 转json对象
	mdconf.mdToJson = function(text) {
		var temp_lines = text.trim().split("\n");
		var lines = [];
		var line = "";
		var conf = {};
		var curConf = conf;

		var getObj = function(key) {
			if (!key) {
				return conf;
			}
			var keys = key.split(".");
			var tmpConf = conf;
			for (var i = 0; i < keys.length; i++){
				tmpConf[keys[i]] = tmpConf[keys[i]] || {};

				if (keys[i].match(/\d+/)) {
					var tmp = parseInt(keys[i]);
					tmpConf.length = tmpConf.length || -1;
					if (tmpConf.length <= tmp) {
						tmpConf.length = tmp + 1;
					}
				
				}
				tmpConf = tmpConf[keys[i]];
			}

			return tmpConf;
		}

		var confConvert = function(c) {
			if (typeof(c) != "object") {
				return c;
			}

			var nc = c.length ? [] : {};

			if (c.length) {
				for (var i = 0; i < c.length; i++) {
					nc.push(confConvert(c[i+""]));
				}
			} else {
				for (var key in c) {
					nc[key] = confConvert(c[key]);
				}
			}

			return nc;
		}

		var _mdToJson = function(line) {
			var temp = line.match(/^([-+#]) (.*)/);
			var flag = temp[1];
			var content = temp[2].trim();	
			var key, value;

			if (flag == "#") {
				curConf = getObj(content);
			}

			if (flag == "+" || flag == "-") {
				temp = content.indexOf(":");

				if (temp > 0) {
					key = content.substring(0, temp).trim(); 
					value = content.substring(temp + 1).trim();
				} else {
					curConf.length = curConf.length || 0;
					key = curConf.length + "";
					value = content.trim();
					curConf.length = curConf.length + 1;
				}

				if (value == "true") {
					value = true;
				} else if (value == "false") {
					value = false;
				} else {
					value = value;
				}

				curConf[key] = value;
			}
		}

		var is_comment = false;
		for (var i = 0; i < temp_lines.length; i++) {
			if (temp_lines[i].match(/^<!--.*-->\s*$/)) {
				continue;
			}
			if (temp_lines[i].match(/^<!--/)) {
				is_comment = true;
				continue;
			}
			if (is_comment) {
				if (temp_lines[i].match(/-->\s*$/)) {
					is_comment = false;
				}
				continue;
			}
			if (!temp_lines[i].match(/^[-+#] .*/)) {
				line += temp_lines[i] + "\n";
				continue;
			}
			if (line) {
				lines.push(line);
			}
			line = temp_lines[i];
		}

		if (line) {
			lines.push(line);
		}

		if (lines.length == 1 && !lines[0].match(/^[-+#] .*/)) {
			return lines[0];
		} else {
			for (var i = 0; i < lines.length; i++) {
				_mdToJson(lines[i]);
			}
		}
		
		return confConvert(conf);
	}

	// json对象转markdown文本
	mdconf.jsonToMd = function(obj) {
		var text = "";
		var value;

		// 非对象直接写入
		if (typeof(obj) != "object") {
			text += obj + "\n";
			return text;
		}
		
		var _jsonToMd = function(obj, key_prefix) {
			key_prefix = key_prefix ? key_prefix + "." : "";
			if (obj.length == undefined) {
				// 单对象对应列表
				for (var key in obj) {
					// 优先写非对象值
					value = obj[key];
					if (key.indexOf("$$") == 0 || typeof(value) == "object") {
						continue;
					}
					text += "- " + key + " : " + value + "\n";
				}
				for (var key in obj) {
					// 写对象值
					value = obj[key];
					if (key.indexOf("$$") == 0 || typeof(value) != "object") {
						continue;
					}

					text += "\n# " + key_prefix + key + "\n";
					_jsonToMd(value, key_prefix + key)
				}
			} else {
				for (var i = 0; i < obj.length; i++) {
					value = obj[i];
					if (typeof(value) != "object") {
						text += "- " + obj[i] + "\n";
						continue;
					} 

					text += "\n# " + key_prefix + i + "\n";
					_jsonToMd(value, key_prefix + i)
				}		
			}
		}

		_jsonToMd(obj);
		return text;
	}

	mdconf.test = function() {
		var obj = undefined, text = undefined;
		//obj = "hello world"
		//text = mdconf.jsonToMd(obj);
		//console.log(text);
		//console.log(angular.toJson(mdconf.mdToJson(text)));
		//console.log(mdconf.mdToJson(text));

		//obj = {key:"value"}
		//text = mdconf.jsonToMd(obj);
		//console.log(text);
		//console.log(angular.toJson(mdconf.mdToJson(text)));
		//console.log(mdconf.mdToJson(text));

		//obj = ["list1", "list2"]
		//text = mdconf.jsonToMd(obj);
		//console.log(text);
		//console.log(angular.toJson(mdconf.mdToJson(text)));
		//console.log(mdconf.mdToJson(text));

		//obj = [["list1", "list2"], ["list3", "list4"]]
		//text = mdconf.jsonToMd(obj);
		//console.log(text);
		//console.log(angular.toJson(mdconf.mdToJson(text)));
		//console.log(mdconf.mdToJson(text));

		//obj = {key:"value", list:["list1", "list2"]}
		//text = mdconf.jsonToMd(obj);
		//console.log(text);
		//console.log(angular.toJson(mdconf.mdToJson(text)));
		//console.log(mdconf.mdToJson(text));

		//obj = {key:"value", list:[{key1:"value1"},{key2:"value2"}]}
		//text = mdconf.jsonToMd(obj);
		//console.log(text);
		//console.log(angular.toJson(mdconf.mdToJson(text)));
		//console.log(mdconf.mdToJson(text));

		//obj = {key:"value", list:[{key1:"value1"},{key2:"value2", list:["list1"]}]}
		//text = mdconf.jsonToMd(obj);
		//console.log(text);
		//console.log(angular.toJson(mdconf.mdToJson(text)));
		//console.log(mdconf.mdToJson(text));

	}

	return mdconf;
})
